tracer/src/Datadog.Trace/ClrProfiler/NativeCallTargetUnmanagedMemoryHelper.cs

```diff
@@ -0,0 +1,403 @@
+﻿// <copyright file="NativeCallTargetUnmanagedMemoryHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Datadog.Trace.ClrProfiler;
+
+internal static class NativeCallTargetUnmanagedMemoryHelper
+{
+    private const int SizeOfMemorySegment = 750 * 1024; // 750Kb
+    private static readonly int SizeOfPointer = Marshal.SizeOf(typeof(IntPtr));
+    private static readonly List<IntPtr> Segments;
+    private static IntPtr _currentSegment;
+    private static int _segmentOffset;
+
+    static NativeCallTargetUnmanagedMemoryHelper()
+    {
+        Segments = new(5);
+        CreateSegment();
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static void CreateSegment()
+    {
+        _currentSegment = Marshal.AllocCoTaskMem(SizeOfMemorySegment);
+        _segmentOffset = 0;
+        Segments.Add(_currentSegment);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static IntPtr Allocate(int bytesCount)
+    {
+        var offset = _segmentOffset;
+        if (SizeOfMemorySegment - offset < bytesCount)
+        {
+            CreateSegment();
```

Comment:
You should also set `offset` to 0, otherwise the logic that follows is wrong

Reply:
True!, done in https://github.com/DataDog/dd-trace-dotnet/pull/4326/commits/74dbe71587ecf6074f3559b5ccad87ee22ec4cfb